nm/bond: Get in-kernel runtime bond options

cathay4t opened the pull request:
NetworkManager has a list of supported bond options in their source
code, use that list and check `/sys/class/net/<bond_name>/bonding` folder
for in-kernel runtime bond options.

When applying bond options, ignore empty string option.

New integration test case added to set all the supported non-default
bond options. Currently with NM 1.22 and 1.20, it might fail due to
https://bugzilla.redhat.com/show_bug.cgi?id=1789437 .

cathay4t commented:
Let's NM feedback on bug https://bugzilla.redhat.com/show_bug.cgi?id=1694966 , in case my test case is incorrect.

tyll commented on libnmstate/nm/bond.py:
```diff
@@ -17,24 +17,59 @@
+SYSFS_BOND_FOLDER = "/sys/class/net/{}/bonding"
+
+NM_SUPPORTED_BOND_OPTIONS = (
+    "mode",
```
Please sort the tuple to make it easier to identify duplicates in case we need to check for new properties in the future.

cathay4t replied:
Sorted. Thanks.

tyll commented on libnmstate/nm/bond.py:
```diff
@@ -17,24 +17,59 @@
 
+SYSFS_BOND_FOLDER = "/sys/class/net/{}/bonding"
+
+NM_SUPPORTED_BOND_OPTIONS = (
```
How often were new properties added in the past? What is the exact scaling problem that you see?

EdDev replied:
I cannot say, I expect not a lot of changes. Sometimes the defaults change.
The scaling problem is about:
- Different bond modes have different properties.
- Additions or removals of options are managed in multiple places (bond kmod, NM, nmstate).

cathay4t replied:
Changed to `frozenset`.

For `Different bond modes have different properties`, the user of nmstate should handle that.

For addition and removals of these options, no good solution neither.

cathay4t replied:
Found a good solution by using NM provided runtime list:

```python
NM_SUPPORTED_BOND_OPTIONS = nmclient.NM.SettingBond.get_valid_options(
    nmclient.NM.SettingBond.new()
)
```

EdDev commented on libnmstate/nm/bond.py:
```diff
@@ -53,12 +88,36 @@ def get_bond_info(nm_device):
+
+    sysfs_folder = SYSFS_BOND_FOLDER.format(ifname)
+    options = {}
+    for sysfs_file_name in os.listdir(sysfs_folder):
```
iglob fits nice here:
```
for option in iglob(f'/sys/class/net/{bond}/bonding/*')
```

EdDev commented on libnmstate/nm/bond.py:
```diff
@@ -53,12 +88,36 @@ def get_bond_info(nm_device):
+    sysfs_folder = SYSFS_BOND_FOLDER.format(ifname)
+    options = {}
+    for sysfs_file_name in os.listdir(sysfs_folder):
+        if sysfs_file_name in NM_SUPPORTED_BOND_OPTIONS:
```
Why this check is needed?

cathay4t replied:
There are many bond options in that folder which is not supported by NM.

EdDev replied:
We can report them all nevertheless, why limit ourself?
When setting, an error will be given I guess.

EdDev replied:
This will also remove the need to maintain the list.

cathay4t commented:
New patch uploaded.

Changes:
 * Sorted the supported bond options.
 * Use `glob.iglob` for iterating the sysfs folder.
 * The value of all bond options should be string. Supporting integer option requires changes to verification stage which should be done in another PR.

EdDev commented on tests/integration/nm/bond_test.py:
```diff
@@ -76,10 +76,23 @@ def _bond_interface(name, options):
-    return _convert_slaves_devices_to_iface_names(nm_bond_info)
+    if not nm_bond_info:
+        return {}
+    state = {
```
nm_bond_info should include this state, why are you re-building it?

cathay4t replied:
We are using assert instead of `assert_state_match()`, we have to remove uninterested bond options.

I will amend the `_bond_interface` context to return a full state.

cathay4t replied:
I don't want to duplicate myself. The same thing has been done in integration test.
I will explain with comment why I rebuild the state.

EdDev commented on tests/lib/nm/bond_test.py:
```diff
@@ -61,19 +61,12 @@ def test_is_bond_type_id(NM_mock):
 
-@mock.patch.object(nm.bond.connection, "ConnectionProfile")
-def test_get_bond_info(con_profile_mock, dev_mock):
+@mock.patch.object(nm.bond, "_get_options")
```
It is improper to mock a private.

cathay4t replied:
Any suggestions?

I don't like change to code only for test purpose.

EdDev replied:
You can mock `iglob` and `open` and then even cover more ground. I think that `open` has a special helper as it is common.

EdDev replied:
BTW, I am also good with dropping this unit test in favor of `nm/bond_test.py`. But in order to "say" that, it is work its own commit with the reasoning in its message.

EdDev replied:
ping

ffmancera commented on libnmstate/nm/bond.py:
```diff
@@ -45,20 +83,38 @@ def is_bond_type_id(type_id):
+    for sysfs_file in glob.iglob(f"/sys/class/net/{ifname}/bonding/*"):
+        option = os.path.basename(sysfs_file)
+        if option in NM_SUPPORTED_BOND_OPTIONS:
+            options[option] = _read_sysfs_file(f"{sysfs_file}")
```
This is generating an options dict with elements of `NM_SUPPORTED_BOND_OPTIONS` on the left side and the values on the right side. Is that okay? Shouldn't we have our API constants on the left instead of `NM_SUPPORTED_BOND_OPTIONS`?

Sorry if I am missing something here.

cathay4t replied:
Don't provide a full list of supported bond options. That's is maintained by NM.

cathay4t commented on libnmstate/nm/bond.py:
```diff
@@ -45,20 +83,38 @@ def is_bond_type_id(type_id):
+    for sysfs_file in glob.iglob(f"/sys/class/net/{ifname}/bonding/*"):
+        option = os.path.basename(sysfs_file)
+        if option in NM_SUPPORTED_BOND_OPTIONS:
+            options[option] = _read_sysfs_file(f"{sysfs_file}")
```
No. The glob.iglob() returns is not a string.

cathay4t replied:
Yeah. Removed.

cathay4t commented on tests/integration/bond_test.py:
```diff
@@ -387,3 +387,79 @@ def test_create_vlan_over_a_bond(bond99_with_slave):
+
+
+parametrize_bond_options = pytest.mark.parametrize(
+    "test_bond_options",
```
Removed these redundant test. Only keep one for known to pass, one for known to fail.

cathay4t commented on tests/integration/bond_test.py:
```diff
@@ -387,3 +387,79 @@ def test_create_vlan_over_a_bond(bond99_with_slave):
+    strict=False,
+)
+@parametrize_bond_options
+def test_change_bond_option(bond99_with_2_slaves, test_bond_options):
```
Removed.

cathay4t commented on libnmstate/nm/bond.py:
```diff
@@ -17,24 +17,59 @@
-            raise NmstateValueError(
-                "Invalid bond option: '{}'='{}'".format(
-                    option_name, option_value
+        if option_value != "":
```
We accept free form of string dict. Why skipping empty value has anything to do with validation?

cathay4t commented:
New patch uploaded.

Changes since previous patch:
 * Use `NM.SettingBond.get_valid_options()` to get the runtime support bond option list.
 * Changed to integration test to only hold one test case for expected pass and one for xfail.

ffmancera commented on libnmstate/nm/bond.py:
```diff
@@ -45,20 +55,38 @@ def is_bond_type_id(type_id):
-    return {}
+def _get_options(nm_device):
+    ifname = nm_device.get_iface()
+    if not ifname:
```
Is that possible? If `nm_device.get_iface()` can be empty at this point maybe we have this issue in other interfaces but I don't think so..

cathay4t replied:
I tried 100 times with bond query after creation/deletion. Never get a empty iface name.
Hence these code removed.

ffmancera replied:
Thanks!

EdDev commented on tests/integration/bond_test.py:
```diff
@@ -394,3 +394,26 @@ def test_create_vlan_over_a_bond(bond99_with_slave):
+    raises=NmstateVerificationError,
+    reason="https://bugzilla.redhat.com/1789437",
+    strict=False,
+)
```
As commented by @thom311 in the mentioned BZ, the `arp_validate` is only relevant for `active-backup` bond mode while this test uses `balance-rr`. The test needs to use a different option.

cathay4t replied:
I have removed this kind of test case.

EdDev commented on tests/integration/nm/bond_test.py:
```diff
@@ -78,9 +78,27 @@ def _bond_interface(name, options):
 
 @nmclient_context
-def _get_bond_current_state(name):
+def _get_bond_current_state(name, option=None):
```
I would have preferred you use plural options, but this is not a must. We can do that when the need will show itself.

EdDev commented on tests/integration/nm/bond_test.py:
```diff
@@ -78,9 +78,27 @@ def _bond_interface(name, options):
+        nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
+            schema.Bond.MODE: bond_options[schema.Bond.MODE],
+            option: bond_options[option],
+        }
```
IMO this way of filtering the data is not a good fit here.
If you just want to assert one or two specific values, change the assert in the test to check only for the relevant parameters or create a dedicated assert function.
Remember that the original asserts fitted well when we did not have support for reading actual properties.

But if you still prefer this way, how about trying to avoid duplicate code here:
```
nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
    schema.Bond.MODE: bond_options[schema.Bond.MODE]
}
if option:
    nm_bond_info[option] = bond_options[option]
```

EdDev commented on tests/lib/nm/bond_test.py:
```diff
@@ -66,19 +66,12 @@ def test_is_bond_type_id(NM_mock):
     expected_info = {
         Bond.SLAVES: dev_mock.get_slaves.return_value,
-        Bond.OPTIONS_SUBTREE: opts_mock,
+        Bond.OPTIONS_SUBTREE: nm.bond._get_options(dev_mock),
```
Same here (private used).

cathay4t replied:
Unit test case removed.

tyll commented:
There seems to be an authentic failure in the CI: https://travis-ci.com/nmstate/nmstate/jobs/288884426#L1454
It looks like the mode is not properly reported:

```
    def test_create_and_remove_bond(eth1_up):
1459        bond_options = {
1460            schema.Bond.MODE: schema.BondMode.ROUND_ROBIN,
1461            "miimon": "140",
1462        }
1463    
1464        with _bond_interface(BOND0, bond_options):
1465            bond_current_state = _get_bond_current_state(BOND0, "miimon")
1466    
1467            bond_desired_state = {
1468                schema.Bond.SLAVES: [],
1469                schema.Bond.OPTIONS_SUBTREE: bond_options,
1470            }
1471>           assert bond_desired_state == bond_current_state
1472E           AssertionError: assert {'options': {... 'slaves': []} == {'options': {'... 'slaves': []}
1473E             Common items:
1474E             {'slaves': []}
1475E             Differing items:
1476E             {'options': {'miimon': '140', 'mode': 'balance-rr'}} != {'options': {'miimon': '140'}}
1477E             Full diff:
1478E             - {'options': {'miimon': '140', 'mode': 'balance-rr'}, 'slaves': []}
1479E             ?                             ----------------------
1480E             + {'options': {'miimon': '140'}, 'slaves': []}
1481
1482tests/integration/nm/bond_test.py:48: AssertionError
```

tyll commented:
@cathay4t 

I added this patch to fix the test failure:

```
diff --git a/tests/integration/nm/bond_test.py b/tests/integration/nm/bond_test.py
index 3f26303..a9a4d69 100644
--- a/tests/integration/nm/bond_test.py
+++ b/tests/integration/nm/bond_test.py
@@ -95,6 +95,7 @@ def _get_bond_current_state(name, option=None):
     }
     if option:
         nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
+            schema.Bond.MODE: bond_options[schema.Bond.MODE],
             option: bond_options[option],
         }
     return _convert_slaves_devices_to_iface_names(nm_bond_info)

```

EdDev commented on tests/integration/nm/bond_test.py:
```diff
@@ -78,9 +78,26 @@ def _bond_interface(name, options):
+        nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
+            schema.Bond.MODE: bond_options[schema.Bond.MODE],
+            option: bond_options[option],
+        }
```
This should be:
```
nm_bond_info[schema.Bond.OPTIONS_SUBTREE][option] = bond_options[option]
```

tyll replied:
why?

EdDev replied:
Because before the condition you got:
```
    nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
        schema.Bond.MODE: bond_options[schema.Bond.MODE]
    }
```
So in the condition body you just need to add the options and nothing more.

tyll commented:
@cathay4t and a follow up change as suggested by @EdDev 

```
diff --git a/tests/integration/nm/bond_test.py b/tests/integration/nm/bond_test.py
index a9a4d69..6c0b1da 100644
--- a/tests/integration/nm/bond_test.py
+++ b/tests/integration/nm/bond_test.py
@@ -94,10 +94,9 @@ def _get_bond_current_state(name, option=None):
         schema.Bond.MODE: bond_options[schema.Bond.MODE]
     }
     if option:
-        nm_bond_info[schema.Bond.OPTIONS_SUBTREE] = {
-            schema.Bond.MODE: bond_options[schema.Bond.MODE],
-            option: bond_options[option],
-        }
+        nm_bond_info[schema.Bond.OPTIONS_SUBTREE][option] = bond_options[
+            option
+        ]
     return _convert_slaves_devices_to_iface_names(nm_bond_info)
 ```

EdDev commented:
The CI failure against NM 1.22.8 is on a known issue. The same passed against the NM stable branch head.
